### PR TITLE
Reread index series rather than storing in memory.

### PR DESCRIPTION
**Pull request by brian-brazil**
Rather than building up a 2nd copy of all the posting
tables, construct it from the data we've already written
to disk. Benchmarks and more details in individual commits.

This saves memory, and also happens to be slightly faster.

### CODE REVIEW

File: tsdb/encoding/encoding.go [hunk 1]
```diff
@@ -260,6 +268,19 @@ func (d *Decbuf) Byte() byte {
 	return x
 }
 
+func (d *Decbuf) EatPadding() {
```
Comment (bwplotka) [inline tsdb/encoding/encoding.go]: Eat? ;p

Comment (brian-brazil) [reply tsdb/encoding/encoding.go]: Do you have a better term?

**Comment by brian-brazil** [comment]
This is ready for review.

File: tsdb/encoding/encoding.go (hunk 1 above)
Comment (Kirides) [reply tsdb/encoding/encoding.go]: `Consume` sounds good :)

Comment (brian-brazil) [reply tsdb/encoding/encoding.go]: Noone for a Devour?

**Comment by brian-brazil** [comment]
So actually I wasn't done with this :)

The performance was good enough compared to the previous implementation, however some quick profiling showed low hanging fruit that means we're overall now ~40% faster than the previous implementation.

File: tsdb/index/index.go [hunk 2]
```diff
@@ -590,76 +594,143 @@ func (w *Writer) writeTOC() error {
...
+		startPos := w.toc.LabelIndices - uint64(d.Len())
+		if startPos%16 != 0 {
+			return errors.Errorf("series not 16-byte aligned at %d", startPos)
 		}
-		if offset > (1<<32)-1 {
-			return errors.Errorf("series offset %d exceeds 4 bytes", offset)
+		offsets = append(offsets, uint32(startPos/16))
+		// Skip to next series. The 4 is for the CRC32.
+		skip := d.Uvarint() + 4
+		d.B = d.B[skip:]
```
Comment (pracucci) [inline tsdb/index/index.go]: We could use `d.Skip()` here too.

File: tsdb/index/index.go [hunk 3]
```diff
@@ -590,76 +594,143 @@ func (w *Writer) writeTOC() error {
...
+	f, err := fileutil.OpenMmapFile(w.f.Name())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Write out the special all index.
+	offsets := []uint32{}
+	d := encoding.NewDecbufRaw(realByteSlice(f.Bytes()), int(w.toc.LabelIndices))
+	d.B = d.B[w.toc.Series:] // dec.Skip not merged yet
```
Comment (pracucci) [inline tsdb/index/index.go]: Now `d.Skip()` is merged and we could use it.

File: tsdb/index/index.go [hunk 4]
```diff
@@ -590,76 +594,143 @@ func (w *Writer) writeTOC() error {
...
-		offset, ok := w.seriesOffsets[it.At()]
-		if !ok {
-			return errors.Errorf("%p series for reference %d not found", w, it.At())
+	f, err := fileutil.OpenMmapFile(w.f.Name())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Write out the special all index.
```
Comment (pracucci) [inline tsdb/index/index.go]: "special all index" sounds unclear to me. Could we improve it?

File: tsdb/index/index.go (hunk 2 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: Though I'd already done that, must have gotten lost. Fixed.

File: tsdb/index/index.go (hunk 4 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: That should say posting.

File: tsdb/encoding/encoding.go [hunk 5]
```diff
@@ -260,6 +268,19 @@ func (d *Decbuf) Byte() byte {
...
+func (d *Decbuf) ConsumePadding() {
+	if d.E != nil {
+		return
+	}
+	for len(d.B) > 1 && d.B[0] == '\x00' {
+		d.B = d.B[1:]
+	}
+	if len(d.B) < 1 {
+		d.E = ErrInvalidSize
+		return
```
Comment (codesome) [inline tsdb/encoding/encoding.go]: Not needed

File: tsdb/index/index.go [hunk 6]
```diff
@@ -590,76 +594,141 @@ func (w *Writer) writeTOC() error {
...
-		if !ok {
-			return errors.Errorf("%p series for reference %d not found", w, it.At())
+	f, err := fileutil.OpenMmapFile(w.f.Name())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Write out the special all posting.
+	offsets := []uint32{}
```
Comment (codesome) [inline tsdb/index/index.go]: If we can keep track of number of series added, we can pre-allocate this and maybe reduce a lot of allocations when there are millions of series.

File: tsdb/index/index.go [hunk 7]
```diff
@@ -590,76 +594,141 @@ func (w *Writer) writeTOC() error {
...
-		return err
-	}
-	w.crc32.Reset()
+	for len(names) > 0 {
+		batchNames := []string{}
+		var c uint64
+		// Try to bunch up label names into one loop, but avoid
+		// using more memory than a single label name can.
+		for len(names) > 0 {
+			if w.labelNames[names[0]]+c > maxPostings {
```
Comment (codesome) [inline tsdb/index/index.go]: `c` always seems to be `0`. Did you miss a `c += w.labelNames[names[0]]` below? If yes, then with this change we might want to run the benchmark again.

File: tsdb/index/index.go [hunk 8]
```diff
@@ -590,76 +594,141 @@ func (w *Writer) writeTOC() error {
...
+	maxPostings := uint64(len(offsets)) // No label name can have more postings than this.
 
-	startPos := w.pos
-	// Leave 4 bytes of space for the length, which will be calculated later.
-	if err := w.write([]byte("alen")); err != nil {
-		return err
-	}
-	w.crc32.Reset()
+	for len(names) > 0 {
+		batchNames := []string{}
```
Comment (codesome) [inline tsdb/index/index.go]: Maybe we could re-use this?

File: tsdb/index/index.go [hunk 9]
```diff
@@ -590,76 +594,141 @@ func (w *Writer) writeTOC() error {
...
+	d.Skip(int(w.toc.Series))
+	for d.Len() > 0 {
+		d.ConsumePadding()
+		startPos := w.toc.LabelIndices - uint64(d.Len())
+		if startPos%16 != 0 {
+			return errors.Errorf("series not 16-byte aligned at %d", startPos)
 		}
-		if offset > (1<<32)-1 {
-			return errors.Errorf("series offset %d exceeds 4 bytes", offset)
+		offsets = append(offsets, uint32(startPos/16))
```
Comment (codesome) [inline tsdb/index/index.go]: Micro-opt: Maybe replace all `/16` with `<<4` if they are done a lot of times.

File: tsdb/index/index.go [hunk 10]
```diff
@@ -590,76 +594,141 @@ func (w *Writer) writeTOC() error {
...
 
-	for _, r := range refs {
-		w.buf1.Reset()
-		w.buf1.PutBE32(r)
-		w.buf1.WriteToHash(w.crc32)
-		if err := w.write(w.buf1.Get()); err != nil {
-			return err
+		for _, name := range batchNames {
+			// Write out postings for this label name.
+			values := make([]uint32, 0, len(postings[w.symbols[name]]))
```
Comment (codesome) [inline tsdb/index/index.go]: We could re-use the `offsets` slice for this.

File: tsdb/index/index.go (hunk 6 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: It'll only be ~20 allocations, not much in the grand scheme of things.

File: tsdb/encoding/encoding.go (hunk 5 above)
Comment (brian-brazil) [reply tsdb/encoding/encoding.go]: Removed, was unsure whether to go for clarity or succinctness here.

File: tsdb/index/index.go (hunk 7 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: Good catch, though it'll make no difference to this particular benchmark as it always gathers up everything.

File: tsdb/index/index.go (hunk 8 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: There's not going to be that many label names. I'd prefer to leave microoptimsation like this to the future.

File: tsdb/index/index.go (hunk 9 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: The compiler should do that already.

File: tsdb/index/index.go (hunk 10 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: Not really, and once again this is only created once per label name.

**Comment by brian-brazil** [comment]
Okay, speeding things up made the compaction cancellation test fail as it's timing dependant. I fixed that, but could someone review the 4th commit before I merge?

**Comment by brian-brazil** [comment]
@bwplotka had a quick peek in IRC, so off we go.

File: tsdb/index/index.go [hunk 11]
```diff
@@ -590,76 +603,147 @@ func (w *Writer) writeTOC() error {
...
-	w.buf1.PutBE32int(int(w.pos - startPos - 4))
-	if err := w.writeAt(w.buf1.Get(), startPos); err != nil {
+func (w *Writer) writePosting(name, value string, offs []uint32) error {
+	// Align beginning to 4 bytes for more efficient postings list scans.
+	if err := w.addPadding(4); err != nil {
 		return err
 	}
 
+	w.postings = append(w.postings, postingsHashEntry{
+		name:   name,
```
Comment (bwplotka) [inline tsdb/index/index.go]: We talked about this offline, but curious if we could store name just once.

**Comment by bwplotka** [comment]
Testing before release if I have permission. 404 so far for me.

File: tsdb/index/index.go (hunk 11 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: That'd be a format change. Getting rid of this line is for a different PR.

File: tsdb/index/index.go [hunk 12]
```diff
@@ -590,76 +594,142 @@ func (w *Writer) writeTOC() error {
...
-		offset, ok := w.seriesOffsets[it.At()]
-		if !ok {
-			return errors.Errorf("%p series for reference %d not found", w, it.At())
+	f, err := fileutil.OpenMmapFile(w.f.Name())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Write out the special all posting.
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/index/index.go]: It's special in the context of the TSDB, as it contains all series in the posting.

File: tsdb/index/index.go [hunk 13]
```diff
@@ -590,76 +603,147 @@ func (w *Writer) writeTOC() error {
...
+			numLabels := d.Uvarint()
+			for i := 0; i < numLabels; i++ {
+				lno := uint32(d.Uvarint())
+				lvo := uint32(d.Uvarint())
+
+				if _, ok := nameSymbols[lno]; ok {
+					if _, ok := postings[lno]; !ok {
+						postings[lno] = map[uint32][]uint32{}
+					}
+					if _, ok := postings[lno][lvo]; !ok {
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/index/index.go]: Removed in next PR.

File: tsdb/index/index.go [hunk 14]
```diff
@@ -590,76 +603,147 @@ func (w *Writer) writeTOC() error {
...
 
-	w.buf1.Reset()
-	w.buf1.PutBE32int(len(refs))
-	w.buf1.WriteToHash(w.crc32)
-	if err := w.write(w.buf1.Get()); err != nil {
-		return err
-	}
+		nameSymbols := map[uint32]struct{}{}
+		for _, name := range batchNames {
+			nameSymbols[w.symbols[name]] = struct{}{}
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/index/index.go]: We could, but I think it's clearer this way.

File: tsdb/index/index.go [hunk 15]
```diff
@@ -590,76 +594,142 @@ func (w *Writer) writeTOC() error {
...
+		d.ConsumePadding()
+		startPos := w.toc.LabelIndices - uint64(d.Len())
+		if startPos%16 != 0 {
+			return errors.Errorf("series not 16-byte aligned at %d", startPos)
 		}
-		if offset > (1<<32)-1 {
-			return errors.Errorf("series offset %d exceeds 4 bytes", offset)
+		offsets = append(offsets, uint32(startPos/16))
+		// Skip to next series. The 4 is for the CRC32.
+		d.Skip(d.Uvarint() + 4)
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/index/index.go]: Added to next PR

**Comment by bwplotka** [comment]
/benchmark cancel

It works for me, as a side note, it's funny as testing on this PR vs master gives a weird difference. Master is without this PR so test makes sense:

image

Some difference in number of chunks 0.o 

image

...and memory (probably because of no chunks):

image

Weird, as we scrape the same data AFAIK .

**Pull request merged**